### ISSUE TITLE
Classify task preprocessing suboptimal for many datasets

### ISSUE THREAD

**Issue by timstokman** [ultralytics/ultralytics#3029]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

The default centercrop preprocessing step for the classify task is suboptimal for many datasets

### Use case

I'm classifying a dataset of people. Something similar to these images:

![person](https://github.com/ultralytics/ultralytics/assets/41363/497943a8-25cf-4e8c-b7e6-8fc0858cee52)

Suppose the dataset consists of images of size 120x640. We're training a model with imgsz 640. The first preprocessing step for the classify task is centercrop:

 ```
class CenterCrop:
    # YOLOv8 CenterCrop class for image preprocessing, i.e. T.Compose([CenterCrop(size), ToTensor()])
    def __init__(self, size=640):
        """Converts an image from numpy array to PyTorch tensor."""
        super().__init__()
        self.h, self.w = (size, size) if isinstance(size, int) else size

    def __call__(self, im):  # im = np.array HWC
        imh, imw = im.shape[:2]
        m = min(imh, imw)  # min dimension
        top, left = (imh - m) // 2, (imw - m) // 2
        return cv2.resize(im[top:top + m, left:left + m], (self.w, self.h), interpolation=cv2.INTER_LINEAR)
```

This _always_ does a center-crop, even if the image fully fits the imgsz, resulting in this:

![center-cropped person](https://github.com/ultralytics/ultralytics/assets/41363/588d3dd6-1467-43c9-802a-99be8171841d)

If you're classifying something like this:

![car](https://github.com/ultralytics/ultralytics/assets/41363/c975d957-6667-4308-87eb-f68a5bc23c71)

A centercrop makes sense, but not implemented like this for all datasets. The torchvision center-crop does not have this issue, it only crops if the image is larger than the set size: https://pytorch.org/vision/main/_modules/torchvision/transforms/functional.html#center_crop

There are some workarounds here (letterbox my own dataset, monkeypatch the augmentations), but this seems like something that should be baked-in.

### Additional

There are several possible solutions for this:

* In `CenterCrop` only crop if the image is too large to fit imgsz
* Use the `ClassifyLetterBox` or `Resize` preprocessing step instead as default
* Ensure the preprocessing step is configurable to some degree (offer options like 'letterbox', 'centercrop', 'resize' etc).

I wouldn't mind submitting a PR for this, but only if I know which option has the preference.

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@timstokman ah interesting! It's always been my goal to eventually unify all preprocessing across tasks, and we took a big step in this direction a few months ago when we removed ImageNet mean and std normalization for classification models.

I was under the assumption that our center crop behaves exactly like Torchvision-trained centercrop classifier models, but if there is a difference in our implementation then it's likely a bug that requires fixing (and retraining). 

Can you please submit a PR for any changes you have in mind, and then we can train a new ImageNet model on the branch to see the effects on accuracy.

**Comment by Laughing-q** [comment]
@timstokman yeah you're right! I think our pretrained model looks good is because many of the images in ImageNet are almost square so `centercrop` does not crop the images that much. But for custom dataset it could be suboptimal since many custom images can be with very high aspect ratio just like your example. So usually I train my custom classify model with `Resize` preprocessing.😂
Hope you can find a nice way to make the preprocessing more generic! Looking forward to the PR!😄

**Comment by timstokman** [comment]
I'll be creating two PRs then:

* One to make the centercrop consistent with torchvision
* One to add some options for the processing step. Centercrop might not always be best for each dataset.

**Comment by glenn-jocher** [comment]
Thank you @timstokman for your contributions and for bringing this issue to our attention. We appreciate your effort in helping us improve YOLOv8!

Your proposed solutions seem like valid approaches to address this issue. Having more options for the preprocessing step could definitely benefit users who may have different requirements for their datasets.

We encourage you to submit your PRs so we can review and test them. In the meantime, if you have any questions or need further assistance, please don't hesitate to ask.

**Comment by timstokman** [comment]
I'll create the second PR once the first gets merged. For the second PR, I was thinking:

* Adding an option 'preprocess' that'll be valid for most tasks. A hyperparameter.
* This option can have the value 'default' (whatever is most appropriate for that model type), 'centercrop', 'letterbox' or 'resize'. Those seem the most relevant options.

I'll have to do a bit of reading in the code-base on how to best do this.

**Comment by glenn-jocher** [comment]
@timstokman thank you for your intention to create a PR to address the issue with the classify task preprocessing. It sounds like you have a thoughtful approach to providing more flexibility in the preprocessing step. 

The addition of an option such as 'preprocess' for the user to select the relevant preprocessing step that works best with their dataset would be a useful addition. Having the option to choose between 'default', 'centercrop', 'letterbox' or 'resize' would provide users with adequate choices based on their requirements. 

We understand that this may require some investigation on your part to implement, but we look forward to seeing your PR and we will review and test it once it's submitted. If you have any questions or need assistance, please don't hesitate to ask.

**Comment by timstokman** [comment]
Since "making the centercrop consistent with the torchvision centercrop" drops the accuracy of the current checkpoints, I'll:
* Close this PR: https://github.com/ultralytics/ultralytics/pull/3050
* Create a PR that:
  * Adds the new suggested preprocessing options
  * Keep the current classification preprocessing setup as default so there's no change to the accuracy of the checkpoint

**Comment by glenn-jocher** [comment]
@timstokman thank you for your effort in addressing the issue with the classify task preprocessing in YOLOv8.

We appreciate the work you put into making the centercrop consistent with torchvision's centercrop. However, as you mentioned, this solution results in a drop in accuracy for the current checkpoints. Given this, we suggest that you close the existing PR and instead create a new PR that adds the preprocessing options you outlined in this thread.

Your proposal of adding an option 'preprocess' that allows the user to select from several preprocessing steps such as 'default', 'centercrop', 'letterbox', or 'resize' sounds like a valuable addition. Keeping the current classification preprocessing setup as the default means there will be no change to the accuracy of the checkpoint.

We look forward to reviewing and testing your PR. Please let us know if you have any questions or need further assistance.

**Comment by timstokman** [comment]
Still trying to get the PR merged, so `bump 👊` so the issue doesn't get closed

**Comment by glenn-jocher** [comment]
@timstokman thank you for your continued effort in getting the PR merged for YOLOv8. We appreciate your dedication and understand the importance of resolving this issue. We will do our best to review and address the PR in a timely manner. If you have any further questions or concerns, please don't hesitate to let us know. Thank you for your patience.

**Comment by XiangjiBU** [comment]
does this problem solved? I got the same problem

**Comment by timstokman** [comment]
I have a pull request that fixes the issue https://github.com/ultralytics/ultralytics/pull/3116, and I am waiting for it to get merged. Needs one of the project's maintainers to look at it and approve it for that to happen.

**Comment by glenn-jocher** [comment]
@XiangjiBU Thank you for your interest in this issue. There is indeed an open pull request that addresses this issue which is currently under review. Once the pull request is approved and merged by the maintainers of the project, this issue should be resolved. Thank you for your patience as we work through this process.

**Comment by timstokman** [comment]
I'm bumping this issue to prevent it from becoming stale. Still trying to get my pull request merged.

**Comment by glenn-jocher** [comment]
Thank you for keeping this on our radar, @timstokman. We understand that your pending pull request is important and appreciate your patience. We aim to offer prompt responses but sometimes it might take longer to thoroughly review code changes. Rest assured, your PR is in our queue and will be attended to as soon as possible by one of our maintainers. Thanks for your contribution to YOLOv8!

**Comment by timstokman** [comment]
I'm bumping this issue to prevent it from becoming stale. Still trying to get my pull request merged, it'd be nice if someone took a look ;).

**Comment by glenn-jocher** [comment]
@timstokman hello, thank you for reaching out and reminding us! We definitely appreciate your effort in addressing this issue through your pull request. Unfortunately, sometimes due to the volume of work it may take longer for reviews to occur. Rest assured, we have noticed your request and it's in our queue. We're all working hard to advance YOLOv8 and your contribution is key to our progress. Thanks for your understanding and patience - we will get to it soon!

**Comment by timstokman** [comment]
Commenting to prevent it from being closed, there's a pull request open

**Comment by glenn-jocher** [comment]
@timstokman your diligence is noted, and the PR will be reviewed. Thank you for your patience. 🛠️
